design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+It is important that we do not rely on hashing or parsing account and region in order to be able to support environment-agnostic stacks (in which case "account" resolves to `{ "Ref": "AWS::AccountId" }`, etc.
+
+In order to address the risk of S3 bucket hijacking, we need to be ale to support an optional `qualifier` postfix. This means that we need to allow users to specify this qualifier when they define the Stack's `env`. Perhaps we need to encode this into `aws://account/region[/qualifier]`
+
+> **Alternative considered**: one way to implement environment-specific name-spacing would have been to export the bootstrapping resources through a CloudFormation Export and then reference them using Fn::ImportValue. This would have worked for templates, but means that we would need a way to resolve import values during publishing as well (and as a result also Fn::Join, etc).
+
+**User Interface**
+
+This is the proposed API for `cdk-bootstrap`:
+
+```
+$ cdk bootstrap --profile XXX [--yes] aws://account/region --trust-account ACCOUNT_ID
+WARNING: any principal from <ACCOUNT_ID> will have administrative access to the account <account>.
+Please confirm (Y/N):
+```
+
+We should consider allowing users to specify a profile map that will allow bootstrapping multiple environments at the same time, but this can easily be achieved through a shell script:
+
+```bash
+#!/bin/sh
+cdk bootstrap --yes --profile prod-us aws://111111111111/us-east-1 --trust-account $PIPELINE_ACCOUNT
+cdk bootstrap --yes --profile prod-eu aws://222222222222/eu-west-2 --trust-account $PIPELINE_ACCOUNT
+```
+
+## Mutation
+
+Deployment of complex cloud application often involves a business-specific process which includes rolling out the app throughout multiple deployment phases and environments. This means that there is strong coupling between the structure of the application and the structure of its deployment pipeline. To enable users to represent this relationship naturally in their code, the CDK should support defining the app's deployment infrastructure as part of the CDK app.
+
+Therefore, we recommend that all resources needed for the deployment pipeline are defined as part of the CDK app itself in one or more stacks. After synthesis, the cloud assembly will include a set of CloudFormation templates for the pipeline stack(s).
+
+We can't begin to deploy an app before we provision and update the required the deployment resources based on the structure of the app. This is the purpose of the "**mutation**" stage.
+
+The initial creation of the pipeline will be performed manually using `cdk deploy pipeline-main` (where `pipeline-main` is name of the main pipeline stack), but from that point forward, any changes to the pipeline will be done by pushing a commit into the repo, and letting the pipeline pick it up.
```

Comment:
This will still cause the Pipeline to stop it's execution after it updates itself (since self modification will cause the current Pipeline execution to fail). 

If you've batched a few changes up, including a Pipeline change, then push - the non-Pipeline changes are going to be orphaned.